[orbitbhyve] Initial contribution

octa22 opened the pull request:
This is the binding for Orbit Bhyve irrrigation systems.

Signed-off-by: Ondrej Pecta <opecta@gmail.com>

kubawolanin commented:
hey @octa22 do you happen to have a JAR for this binding for latest milestone build?
I do have a B-Hyve indoor sprinkler system and I'm willing to test your work on a real device if ready :)
Cheers

octa22 commented:
Hi, here is the latest 3.2.0 SNAPSHOT. I'd be happy if you can test it.
https://www.dropbox.com/s/hpdxn1cjtn9xj0v/org.openhab.binding.orbitbhyve-3.2.0-SNAPSHOT.jar?dl=1
Ondrej

kubawolanin commented:
Hi @octa22!

I've placed the JAR in my addons folder and added a Bridge manually.
For some reason the discovery did not kick in.

After adding a Bridge and immediately after a Sprinkler, I realized that I had a typo in the password field.
This resulted in the following errors after authentication:
```
 events.log - 2021-06-29 16:03:42.587 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'orbitbhyve:bridge:sprinklers' changed from OFFLINE (COMMUNICATION_ERROR) to OFFLINE (COMMUNICATION_ERROR): Cannot create websocket session
 events.log - 2021-06-29 16:03:42.587 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'orbitbhyve:bridge:sprinklers' changed from OFFLINE (COMMUNICATION_ERROR): Cannot create websocket session to ONLINE
 events.log - 2021-06-29 16:03:42.587 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'orbitbhyve:sprinkler:sprinklers:ddcd8f8b63' changed from OFFLINE (BRIDGE_OFFLINE) to ONLINE
 events.log - 2021-06-29 16:03:43.334 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'orbitbhyve:bridge:sprinklers' changed from ONLINE to OFFLINE (COMMUNICATION_ERROR): Cannot create websocket session
 events.log - 2021-06-29 16:03:43.334 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'orbitbhyve:bridge:sprinklers' changed from OFFLINE (COMMUNICATION_ERROR): Cannot create websocket session to OFFLINE (COMMUNICATION_ERROR): web socket creation error
 events.log - 2021-06-29 16:03:43.334 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'orbitbhyve:bridge:sprinklers' changed from OFFLINE (COMMUNICATION_ERROR): web socket creation error to OFFLINE (COMMUNICATION_ERROR): Returned status: 400
 events.log - 2021-06-29 16:03:43.506 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'orbitbhyve:bridge:sprinklers' changed from OFFLINE (COMMUNICATION_ERROR): Returned status: 400 to OFFLINE
 events.log - 2021-06-29 16:03:43.506 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'orbitbhyve:sprinkler:sprinklers:ddcd8f8b63' changed from ONLINE to OFFLINE (BRIDGE_OFFLINE)
```

Then every 30 seconds it returns 400 HTTP status:

```
 events.log - 2021-06-29 16:05:17.203 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'orbitbhyve:bridge:sprinklers' changed from OFFLINE (COMMUNICATION_ERROR): Returned status: 400 to ONLINE
 events.log - 2021-06-29 16:05:17.203 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'orbitbhyve:sprinkler:sprinklers:ddcd8f8b63' changed from OFFLINE (BRIDGE_OFFLINE) to ONLINE
 events.log - 2021-06-29 16:05:17.378 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'orbitbhyve:bridge:sprinklers' changed from ONLINE to OFFLINE (COMMUNICATION_ERROR): Returned status: 400
 events.log - 2021-06-29 16:05:17.746 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'orbitbhyve:sprinkler:sprinklers:ddcd8f8b63' changed from ONLINE to OFFLINE (BRIDGE_OFFLINE)
```

What I did after that was I completely removed all the created things and started over - this time making sure that my password was correct before I hit Save in the UI during Bridge thing creation.
Turned out to be successful:
![image](https://user-images.githubusercontent.com/2270505/123812423-5e42fb80-d8f4-11eb-9908-679108c14f92.png)

After I waited for the bridge to become Online, I navigated to Inbox and then saw that discovery indeed works:
![image](https://user-images.githubusercontent.com/2270505/123812561-79ae0680-d8f4-11eb-9826-7db12bbaf5a6.png)


```
 events.log - 2021-06-29 16:10:20.746 [INFO ] [openhab.event.InboxRemovedEvent     ] - Discovery Result with UID 'orbitbhyve:sprinkler:16f92e4ed5:5f7759994f0c62a04bd24722' has been removed.
 events.log - 2021-06-29 16:10:20.747 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'orbitbhyve:sprinkler:16f92e4ed5:5f7759994f0c62a04bd24722' changed from UNINITIALIZED to INITIALIZING
 events.log - 2021-06-29 16:10:21.445 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'orbitbhyve:sprinkler:16f92e4ed5:5f7759994f0c62a04bd24722' changed from INITIALIZING to ONLINE
```

Starting Smart Watering works perfectly:
```
 events.log - 2021-06-29 16:11:54.528 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'Sprinklers_IrrigationMode' changed from NULL to auto
 events.log - 2021-06-29 16:11:54.528 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'Sprinklers_SprinklerStateControl' changed from NULL to ON
 events.log - 2021-06-29 16:11:54.529 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'Sprinklers_NextWateringStartTime' changed from NULL to 1970-01-01T00:00:00.000-0400
```

Thank you Ondrej for your great work! 🎉

kubawolanin commented:
I've added Equipment to Model with the UI but noticed that it doesn't update Items state:
![image](https://user-images.githubusercontent.com/2270505/123837937-b1747880-d90b-11eb-9d21-12a9873d9e2a.png)

And unfortunately none of the switches seem to work :(

octa22 commented:
Hi,

the binding should automatically detect all programs and zones you have (as
switches where the ON command enables the watering or the program)
This is the way how to enable manual watering.

Apparently the is some error during the communication with the B-Hybe cloud.
It is working fine for the 8-zones Orbit sprinkler that I have.

Please can you enable the trace log and provide it here?
Thanks.
Ondrej

út 29. 6. 2021 v 18:56 odesílatel Kuba Wolanin ***@***.***>
napsal:

> I've added Equipment to Model with the UI but noticed that it doesn't
> update Items state:
> [image: image]
> <https://user-images.githubusercontent.com/2270505/123837937-b1747880-d90b-11eb-9d21-12a9873d9e2a.png>
>
> And unfortunately none of the switches seem to work :(
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openhab/openhab-addons/pull/10426#issuecomment-870762191>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE7QO52K2NTLZOEH76GWH7TTVH3NRANCNFSM42HW5YQQ>
> .
>

octa22 commented:
Hello @kubawolanin,
indeed there were some compatibility issues with the OH3.
It should be fixed now. Please, can you download the jar again and let me know?
https://www.dropbox.com/s/hpdxn1cjtn9xj0v/org.openhab.binding.orbitbhyve-3.2.0-SNAPSHOT.jar?dl=1
All zones and programs should appear as new channels and you will be able to trigger the watering (a zone or a program)
Thanks.
Ondrej

kubawolanin commented:
Hi @octa22 

Re-added the things after installing the latest binding and it immediately found all my zones!
![image](https://user-images.githubusercontent.com/2270505/124380249-07567100-dcbc-11eb-9e3b-2653ac318230.png)

Manual watering works perfectly - here's a proof ;-)

![image](https://user-images.githubusercontent.com/2270505/124380311-4a184900-dcbc-11eb-9f6a-62d3a29d5f1c.png)
![image](https://user-images.githubusercontent.com/2270505/124380293-3e2c8700-dcbc-11eb-9fd3-27e44c8fb56a.png)

Unfortunately I forgot to switch to TRACE when adding things but will test the binding for a few days.

Thank you for the great work!

octa22 commented:
Hi @kubawolanin, thanks again for the testing. Very glad it works! May I ask what device are you using? I am finishing the documentation and I would like to include a list of tested devices so far.
Thank you!

kubawolanin commented:
It's B-Hyve Smart Indoor Timer @octa22

octa22 commented:
> It's B-Hyve Smart Indoor Timer @octa22 

Thanks. The same as mine. I guess the binding is ready for a review. @kubawolanin

fwolter commented on bundles/org.openhab.binding.orbitbhyve/src/main/resources/OH-INF/binding/binding.xml:
```diff
@@ -0,0 +1,9 @@
+	xsi:schemaLocation="https://openhab.org/schemas/binding/v1.0.0 https://openhab.org/schemas/binding-1.0.0.xsd">
+
+	<name>Orbit B-Hyve Binding</name>
+	<description>This is the binding for Orbit B-Hyve.</description>
```
Could you be a bit more verbose, so that somebody who hasn't heard about it gets a clue what this is about?

octa22 replied:
Fixed

fwolter commented on bundles/org.openhab.binding.orbitbhyve/src/main/feature/feature.xml:
```diff
@@ -0,0 +1,9 @@
+<features name="org.openhab.binding.orbitbhyve-${project.version}" xmlns="http://karaf.apache.org/xmlns/features/v1.4.0">
+	<repository>mvn:org.openhab.core.features.karaf/org.openhab.core.features.karaf.openhab-core/${ohc.version}/xml/features</repository>
+
+	<feature name="openhab-binding-orbitbhyve" description="OrbitBhyve Binding" version="${project.version}">
```
This name will be displayed, when installing the binding via the UI.
```suggestion
	<feature name="openhab-binding-orbitbhyve" description="Orbit B-Hyve Binding" version="${project.version}">
```

octa22 replied:
Fixed

fwolter commented on bundles/org.openhab.binding.orbitbhyve/pom.xml:
```diff
@@ -0,0 +1,17 @@
+
+  <artifactId>org.openhab.binding.orbitbhyve</artifactId>
+
+  <name>openHAB Add-ons :: Bundles :: OrbitBhyve Binding</name>
```
```suggestion
  <name>openHAB Add-ons :: Bundles :: Orbit B-Hyve Binding</name>
```

octa22 replied:
Fixed

fwolter commented on bundles/org.openhab.binding.orbitbhyve/src/main/resources/OH-INF/config/config.xml:
```diff
@@ -0,0 +1,30 @@
+
+	<config-description uri="bridge-type:orbitbhyve:bridge">
+		<parameter name="email" type="text" required="true">
+			<label>Email parameter</label>
```
```suggestion
			<label>Email</label>
```

octa22 replied:
made it shorter

fwolter commented on bundles/org.openhab.binding.orbitbhyve/src/main/resources/OH-INF/config/config.xml:
```diff
@@ -0,0 +1,30 @@
+
+	<config-description uri="bridge-type:orbitbhyve:bridge">
+		<parameter name="email" type="text" required="true">
+			<label>Email parameter</label>
```
You could specify `<context>email</context>` to get a free validation.

octa22 replied:
added context

fwolter commented on bundles/org.openhab.binding.orbitbhyve/src/main/resources/OH-INF/thing/bridge.xml:
```diff
@@ -0,0 +1,13 @@
+
+	<!-- Bridge Type -->
+	<bridge-type id="bridge">
+		<label>Orbit B-Hyve Binding Bridge</label>
```
Labels are expected to be as short as possible. Guideline is 2-3 words with max 20-25 chars. See https://www.openhab.org/docs/developer/bindings/thing-xml.html#formatting-labels-and-descriptions

Also, the binding name doesn't need to be repeated in the Thing label.

octa22 replied:
Fixed

fwolter commented on bundles/org.openhab.binding.orbitbhyve/src/main/resources/OH-INF/thing/channels.xml:
```diff
@@ -0,0 +1,52 @@
+		<item-type>String</item-type>
+		<label>Irrigation Mode</label>
+		<description>Channel representing mode of Orbit B-Hyve Device (auto/manual)</description>
+		<state readOnly="true"></state>
```
You could specify the two options within the state tag.

octa22 replied:
added both suggestions

fwolter commented on bundles/org.openhab.binding.orbitbhyve/src/main/resources/OH-INF/thing/channels.xml:
```diff
@@ -0,0 +1,52 @@
+	</channel-type>
+	<channel-type id="rain_delay">
+		<item-type>Integer</item-type>
+		<label>Rain delay</label>
```
Words in labels should be capitalized (except prepositions and so on). See https://www.openhab.org/docs/developer/bindings/thing-xml.html#formatting-labels-and-descriptions

octa22 replied:
Fixed label capitalisation (more occurences)

fwolter commented on bundles/org.openhab.binding.orbitbhyve/src/main/resources/OH-INF/thing/channels.xml:
```diff
@@ -0,0 +1,52 @@
+		<state readOnly="true"></state>
+	</channel-type>
+	<channel-type id="rain_delay">
+		<item-type>Integer</item-type>
```
Please check all.
```suggestion
		<item-type>Number:Time</item-type>
```

octa22 replied:
Removed Integer and replaced by Number:Time and Number

fwolter commented on bundles/org.openhab.binding.orbitbhyve/src/main/resources/OH-INF/thing/channels.xml:
```diff
@@ -0,0 +1,52 @@
+	</channel-type>
+	<channel-type id="smart_watering">
+		<item-type>Switch</item-type>
+		<label>Sprinkler Smart Watering Control</label>
```
Labels are expected to be as short as possible. Guideline is 2-3 words with max 20-25 chars. See https://www.openhab.org/docs/developer/bindings/thing-xml.html#formatting-labels-and-descriptions

octa22 replied:
Made it shorter

fwolter commented on bundles/org.openhab.binding.orbitbhyve/src/main/resources/OH-INF/thing/sprinkler.xml:
```diff
@@ -0,0 +1,23 @@
+		<supported-bridge-type-refs>
+			<bridge-type-ref id="bridge"/>
+		</supported-bridge-type-refs>
+		<label>Orbit B-Hyve Sprinkler</label>
```
```suggestion
		<label>Sprinkler</label>
```

octa22 replied:
Fixed

fwolter reviewed (changes requested):
Thanks for your contribution! I reviewed your code and here is my feedback.

You need to add your binding to `bom\openhab-addons\pom.xml`.

Your binding doesn't build as it contains checkstyle errors. You could take a look at `target/code-analysis/report.html` after building it with `mvn clean install`.

There is also one compiler warning about a null value left.

octa22 commented on bundles/org.openhab.binding.orbitbhyve/README.md:
```diff
@@ -0,0 +1,90 @@
+
+## Binding Configuration
+
+This binding does not require any configuration on the binding level.
```
removed

octa22 commented on bundles/org.openhab.binding.orbitbhyve/README.md:
```diff
@@ -0,0 +1,90 @@
+| mode           | String   | This channel represents the mode of sprinkler device (auto/manual) |
+| next_start     | DateTime | This channel represents the start time of the next watering        |
+| rain_delay     | Number   | This channel manages the current rain delay in hours               |
+| watering_time  | Number   | This channel manages the manual zone watering time in minutes      |
```
changed to Number:Time

octa22 commented on bundles/org.openhab.binding.orbitbhyve/README.md:
```diff
@@ -0,0 +1,90 @@
+| watering_time  | Number   | This channel manages the manual zone watering time in minutes      |
+| control        | Switch   | This channel controls the sprinkler (ON/OFF)                       |
+| smart_watering | Switch   | This channel controls the smart watering (ON/OFF)                  |
+| control        | String   | This channel manages the manual zone watering time in minutes      |
```
Sorry - it is a Copy & Paste relict, removed

octa22 commented on bundles/org.openhab.binding.orbitbhyve/src/main/java/org/openhab/binding/orbitbhyve/internal/OrbitBhyveHandlerFactory.java:
```diff
@@ -0,0 +1,87 @@
+    /**
+     * the shared http client
+     */
+    private @NonNullByDefault({}) HttpClient httpClient;
```
removed the annotation

octa22 commented on bundles/org.openhab.binding.orbitbhyve/src/main/java/org/openhab/binding/orbitbhyve/internal/handler/OrbitBhyveBridgeHandler.java:
```diff
@@ -0,0 +1,588 @@
+    @Override
+    public void dispose() {
+        ScheduledFuture<?> localFuture = future;
+        if (localFuture != null && !(localFuture.isCancelled() || localFuture.isDone())) {
```
Fixed

octa22 commented:
Hi @fwolter,
thanks for the review. I guess I have addressed all your comments.
Checkstyle errors have been fixed and the I have added the entry to `bom\openhab-addons\pom.xml`

octa22 commented on bundles/org.openhab.binding.orbitbhyve/src/main/java/org/openhab/binding/orbitbhyve/internal/handler/OrbitBhyveBridgeHandler.java:
```diff
@@ -0,0 +1,588 @@
+                return false;
+            }
+        } catch (InterruptedException | TimeoutException | ExecutionException e) {
+            logger.debug("Exception during login", e);
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR);
```
Fixed

octa22 commented on bundles/org.openhab.binding.orbitbhyve/src/main/java/org/openhab/binding/orbitbhyve/internal/handler/OrbitBhyveBridgeHandler.java:
```diff
@@ -0,0 +1,588 @@
+                    logger.debug("Sending ping");
+                    localSession.getRemote().sendString("{\"event\":\"ping\"}");
+                } catch (IOException e) {
+                    logger.debug("Error sending ping to a web socket", e);
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                            "web socket communication error");
```
Logging removed

octa22 commented on bundles/org.openhab.binding.orbitbhyve/src/main/java/org/openhab/binding/orbitbhyve/internal/handler/OrbitBhyveBridgeHandler.java:
```diff
@@ -0,0 +1,588 @@
+                logger.debug("token: {}", sessionToken);
+                initializeWebSocketSession();
+            } else {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR);
```
Added response status code as a status detail message

octa22 commented on bundles/org.openhab.binding.orbitbhyve/src/main/java/org/openhab/binding/orbitbhyve/internal/handler/OrbitBhyveBridgeHandler.java:
```diff
@@ -0,0 +1,588 @@
+                if (ch != null) {
+                    updateState(ch.getUID(), "off".equals(event.getMode()) ? OnOffType.OFF : OnOffType.ON);
+                }
+                scheduler.schedule(runnableUpdateDeviceStatus, 3, TimeUnit.SECONDS);
```
I have found out that there is a slight delay between the status change is taken into account. So I need a delay before querying the updated status. I have changed the way of deferring the status updating and clean disposing as well.

fwolter commented on bundles/org.openhab.binding.orbitbhyve/README.md:
```diff
@@ -0,0 +1,84 @@
+|----------------|-------------|--------------------------------------------------------------------|
+| mode           | String      | This channel represents the mode of sprinkler device (auto/manual) |
+| next_start     | DateTime    | This channel represents the start time of the next watering        |
+| rain_delay     | Number      | This channel manages the current rain delay in hours               |
```
Seems this can be Number:Time, too.

octa22 replied:
fixed

fwolter commented on bundles/org.openhab.binding.orbitbhyve/src/main/java/org/openhab/binding/orbitbhyve/internal/OrbitBhyveHandlerFactory.java:
```diff
@@ -0,0 +1,87 @@
+@Component(configurationPid = "binding.orbitbhyve", service = ThingHandlerFactory.class)
+public class OrbitBhyveHandlerFactory extends BaseThingHandlerFactory {
+
+    private static final Set<ThingTypeUID> SUPPORTED_THING_TYPES_UIDS = new HashSet<>(
+            Arrays.asList(THING_TYPE_BRIDGE, THING_TYPE_SPRINKLER));
```
```suggestion
    private static final Set<ThingTypeUID> SUPPORTED_THING_TYPES_UIDS = Set.of(THING_TYPE_BRIDGE, THING_TYPE_SPRINKLER);
```

octa22 replied:
done

fwolter commented on bundles/org.openhab.binding.orbitbhyve/src/main/java/org/openhab/binding/orbitbhyve/internal/OrbitBhyveHandlerFactory.java:
```diff
@@ -0,0 +1,87 @@
+    /**
+     * the shared web socket client
+     */
+    private @NonNullByDefault({}) WebSocketClient webSocketClient;
```
This null annotation can be removed, too.

fwolter commented on bundles/org.openhab.binding.orbitbhyve/src/main/java/org/openhab/binding/orbitbhyve/internal/handler/OrbitBhyveBridgeHandler.java:
```diff
@@ -0,0 +1,584 @@
+        }
+    }
+
+    private void deferredStatusUpdate(String deviceId) {
+        ScheduledFuture<?> localFuture = statusFuture;
+        if (localFuture != null) {
+            localFuture.cancel(true);
+        }
+        statusFuture = scheduler.schedule(() -> updateDeviceStatus(deviceId), 3, TimeUnit.SECONDS);
```
This won't work if the user sends two commands within less than 3 seconds, as `statusFuture` will be overwritten, then.

Are you sure the update is needed at all? Actually, openHAB predicts the new state automatically when a command is sent to a channel. See https://www.openhab.org/docs/developer/bindings/thing-xml.html#auto-update-policies

octa22 replied:
I have changed the behaviour so it call the device status immediately and ping method does status reconciliation, so any delayed status updates are taken into account.

fwolter replied:
Actually, you don't need to request the new state from the device as the framework predicts the new state immediately and verifies it during the next polling. But it's ok to keep it for now.

fwolter commented on bundles/org.openhab.binding.orbitbhyve/src/main/java/org/openhab/binding/orbitbhyve/internal/handler/OrbitBhyveSprinklerHandler.java:
```diff
@@ -0,0 +1,261 @@
+                handler.stopWatering(deviceId);
+                return;
+            }
+            if (CHANNEL_WATERING_TIME.equals(channelUID.getId()) && command instanceof DecimalType) {
+                wateringTime = ((DecimalType) command).intValue();
+                updateState(CHANNEL_WATERING_TIME, (DecimalType) command);
+                return;
```
If you make this Number:Time, the command is of type `QuantityType`. As the user can send a command with any time unit, you need to convert the value to the the unit you expect with `toUnit()`. Take care that the return value can be null, if the unit cannot be converted.

octa22 replied:
ok, fixed

fwolter commented on bundles/org.openhab.binding.orbitbhyve/src/main/java/org/openhab/binding/orbitbhyve/internal/OrbitBhyveHandlerFactory.java:
```diff
@@ -0,0 +1,86 @@
+@Component(configurationPid = "binding.orbitbhyve", service = ThingHandlerFactory.class)
+public class OrbitBhyveHandlerFactory extends BaseThingHandlerFactory {
+
+    private static final Set<ThingTypeUID> SUPPORTED_THING_TYPES_UIDS = new HashSet<>(
+            Set.of(THING_TYPE_BRIDGE, THING_TYPE_SPRINKLER));
```
```suggestion
    private static final Set<ThingTypeUID> SUPPORTED_THING_TYPES_UIDS = Set.of(THING_TYPE_BRIDGE, THING_TYPE_SPRINKLER);
```

fwolter commented:
Congratulation! Now, you could add your binding's logo to the openHAB website. See https://next.openhab.org/docs/developer/bindings/#add-your-binding-s-logo-to-the-openhab-website
